feat: Support NestJS `filter` function for subscriptions

gthau opened the pull request:
NestJS subscriptions' resolvers have a `filter` and a `resolve` function in the @Subscription decorator. Currently only the `resolve` is supported by the YogaDriver for NestJS.

This commit adds support for the `filter` method. The implementation is similar to NestJS Apollo plugin, but it uses Yoga's own async iterator utils (pipe, filter) instead of Apollo's utils.

n1ru4l commented:
This can be achieved with function composition using the filter and pipe operators exported from yoga. I see no need to add this in the API.

bernessco commented:
Hello!
I suggest reading through https://github.com/dotansimha/graphql-yoga/issues/2978
In short, the subscription decorator has a filter method that is used in other drivers and if you want to switch to yoga you will need to rewrite the code which is not ideal. For my example migrating 200+ subscriptions. Another point is that nest documentation guides you to use filters in decorator which can lead to data leaks because the filter argument is ignored and never called in the yoga driver. This is what happened to me, I did not notice that the filters stopped working and stared pushing all events to frontend

enisdenjo commented:
Thank you @gthau for this great PR, it looks awesome! I would kindly ask you to run `pnpm changeset` and create a new minor release for this feature, thanks in advance.

Also thanks to @bernessco for elaborating the need, very insightful.

gthau commented:
Thank you @bernessco  and @enisdenjo!
Always a pleasure collaborating with The Guild.

enisdenjo commented:
This is now live in [`@graphql-yoga/nestjs@2.1.0` and `@graphql-yoga/nestjs-federation@2.2.0`](https://github.com/dotansimha/graphql-yoga/releases/tag/release-1693921958077). Feedback is very welcome!

gthau commented:
> This is now live in [`@graphql-yoga/nestjs@2.1.0` and `@graphql-yoga/nestjs-federation@2.2.0`](https://github.com/dotansimha/graphql-yoga/releases/tag/release-1693921958077). Feedback is very welcome!

I tested on my Nest project, updated  with `@graphql-yoga/nestjs@2.1.0`, removed my custom PatchedYogaDriver (which implemented `subscriptionWithFilter`), it works as expected.

I can't give feedback about `@graphql-yoga/nestjs-federation` as I don't use it.
